MobiFlight/Joysticks/Octavi/OctaviHandler.cs

```diff
@@ -393,47 +320,73 @@ public OctaviHandler()
+            else if (report.outerEncoderDelta < 0)
+            {
+                for (int i = 0; i > report.outerEncoderDelta; i--)
+                {
+                    buttonPresses.Add((OctaviButtonMatrix.ElementAt(extendedContextState * 16 + 1).Value, MobiFlightButton.InputEvent.PRESS));
+                }
+            }
+            if (report.innerEncoderDelta > 0)
+            {
+                for (int i = 0; i < report.innerEncoderDelta; i++)
+                {
+                    buttonPresses.Add((OctaviButtonMatrix.ElementAt(extendedContextState * 16 + 2).Value, MobiFlightButton.InputEvent.PRESS));
+                }
+            }
+            else if (report.innerEncoderDelta < 0)
+            {
+                for (int i = 0; i > report.innerEncoderDelta; i--)
+                {
+                    buttonPresses.Add((OctaviButtonMatrix.ElementAt(extendedContextState * 16 + 3).Value, MobiFlightButton.InputEvent.PRESS));
+                }
+            }
 
-            if((pressed & (uint)OctaviReport.OctaviButton.HID_ENC_SW)!=0)
+            if (pressed.HasFlag(OctaviReport.OctaviButtons.HID_ENC_SW))
             {
-                //if (report.contextState != (byte)OctaviReport.OctaviState.STATE_FMS1 && report.contextState != (byte)OctaviReport.OctaviState.STATE_FMS2 && report.contextState != (byte)OctaviReport.OctaviState.STATE_NAV1 && report.contextState != (byte)OctaviReport.OctaviState.STATE_NAV2)
-                if (report.contextState != (byte)OctaviReport.OctaviState.STATE_FMS1 && report.contextState != (byte)OctaviReport.OctaviState.STATE_FMS2)
+                if (report.contextState != OctaviReport.OctaviState.STATE_FMS1 && report.contextState != OctaviReport.OctaviState.STATE_FMS2)
                 {
-                        shiftState = !shiftState; // FMS1&2 do not have shift modes for now, sorry
+                    isInShiftMode = !isInShiftMode; // FMS1&2 do not have shift modes for now, sorry
                 }
             }
-            foreach(uint HIDEvent in HIDEventAssignments.Keys)
+            foreach (OctaviReport.OctaviButtons button in Enum.GetValues(typeof(OctaviReport.OctaviButtons)))
             {
-                if ((pressed & HIDEvent) != 0)
+                if (pressed.HasFlag(button) || released.HasFlag(button))
                 {
-                    int ButtonPress = extContentState * 16 + (int)HIDEventAssignments[HIDEvent]; // find "full matrix" event index
-                    ButtonPress = OctaviButtonMatrix.ElementAt(ButtonPress).Value; // translate to existing Octavi "devices" in MF
-                    if (ButtonPress >= 0) ButtonPresses.Add(ButtonPress); // if not "unassigned" (-1), then press the button!
+                    int lookupIndex = extendedContextState * 16 + HidEventAssignments[button]; // find "full matrix" event index
+                    int buttonIndex = OctaviButtonMatrix.ElementAt(lookupIndex).Value; // translate to existing Octavi "devices" in MF
```

Comment:
OctaviButtonMatrix is a Dictionary which has no guaranteed order. Using ElementAt is not really the correct way to do this.